### ISSUE TITLE
Forward pass with dropout

### ISSUE THREAD

**Issue by beamandrew** [fchollet/keras#1606]
I'm looking for a way to do something that seems conceptually simple, but I haven't found a way to do it in Keras yet. I have a CNN with dropout applied in various places in the network. After training, I'd like to compute the forward pass on some data and have the dropout mask applied such that each time I compute the forward pass a different dropout mask is applied. 

If I understand things correctly, this is what is done during training. However, if I use the standard model.predict(X_test) after training, no dropout is applied and the outputs are multiplied by the dropout probability, yielding a non-stochastic set of predictions. 

Does anyone know how I can achieve this?

**Comment by fchollet** [comment]
Permanent dropout:
```python
from keras.layers.core import Lambda
from keras import backend as K

model.add(Lambda(lambda x: K.dropout(x, level=0.5)))
```

**Comment by cbonnett** [comment]
There is no elegant way of doing this with a 'normally' defined  model that at test time can be evaluated with and without the dropout mask ?

**Comment by fchollet** [comment]
The simplest thing you can do is use your own dropout layer. Look at the
code for the existing dropout layer. Simply remove the part that makes
dropout conditional on the learning phase.

**Comment by iskandr** [comment]
@fchollet How do we generalize this answer to other forms of dropout, such as those in Embedding and RNN layers?

**Comment by dpgfoures** [comment]
```
class PermanentDropout(Dropout):
    def __init__(self, rate, **kwargs):
        super(PermanentDropout, self).__init__(rate, **kwargs)
        self.uses_learning_phase = False

    def call(self, x, mask=None):
        if 0. < self.rate < 1.:
            noise_shape = self._get_noise_shape(x)
            x = K.dropout(x, self.rate, noise_shape)
        return x
```

**Comment by jordanott** [comment]
@fchollet I am interested in making a dropout layer that is static throughout the course of training and testing. Is creating a custom layer the easiest way to achieve this? It seems I would need to modify/write additional tensorflow source code to do this, specifically [here](https://github.com/tensorflow/tensorflow/blob/4acb96a9708a7d791e5d005e22971a05b56adcbd/tensorflow/python/ops/nn_ops.py#L1861) .

**Comment by andrisecker** [comment]
@fchollet is there a way to do the same with spatial dropout?

**Comment by sgfin** [comment]
Based on issue #9412, it appears that this "permanent dropout" feature has been (quietly) added into core Keras.

Per @fchollet:

> There is this feature in Keras: it's the training argument in the call of the Dropout layer.
> 
> Here's a model with a Dense layer and a Dropout layer that runs both in training and testing:

`import keras

inputs = keras.Input(shape=(10,))
x = keras.layers.Dense(3)(inputs)
outputs = keras.layers.Dropout(0.5)(x, training=True)

model = keras.Model(inputs, outputs)`

Oddly, the `training` argument doesn't currently appeared to be documented on [https://keras.io/layers/core/](https://keras.io/layers/core/), but you can see it in line 118 of the [source](https://github.com/keras-team/keras/blob/master/keras/layers/core.py#L80).

**Comment by cerlymarco** [comment]
You can also change the dropout with a trained model:

`f = K.function([model.layers[0].input, K.learning_phase()],
               [model.layers[-1].output])`

In this way you don't have to train again the model!!!

**Comment by birajaghoshal** [comment]
How can I pass dropout rate during predict in Keras?
for example:
dropout_rate = 0.1
prediction = model.predict(x_test, dropout_rate)

**Comment by milind-prajapat** [comment]
> You can also change the dropout with a trained model:
> 
> `f = K.function([model.layers[0].input, K.learning_phase()], [model.layers[-1].output])`
> 
> In this way you don't have to train again the model!!!

This is throwing error.

**Comment by DavidWAbrahams** [comment]
This is what my minimum viable solution looks like in 2022:
`model(input_data, training=True)`

That works with the sequential API at least.

That lets me run a model with dropout multiple times and generate a sort of confidence score:
```
tolerance = <how much prediction error you can tolerate>
num_iterations = 50
prediction = your_model_with_dropout(input_data)  # doesn't trigger dropout
# run inference on the same data point multiple times, with dropout
test_predictions = your_model_with_dropout(np.tile(input_data, (num_iterations, 1)), training=True).numpy()
lower_bound = prediction - tolerance
upper_bound = prediction + tolerance
# What fraction of the predictions were within the allowed tolerance?
confidence = np.count_nonzero((test_predictions > lower_bound ) & (test_predictions < upper_bound )) / len(test_predictions )
```
